File: 9.DetectingCycle.py
#Time O(n) space O(n)
def detectingCycle(head):
    seen = []
    currNode = head
    
    while currNode not in seen:
        seen.append(currNode)
        if currNode and currNode.next:
            currNode = currNode.next
        else:
            return False
        
        if currNode in seen:
            return True
        
    return False


#Time O(n) space O(1)
def tortoiseAndHare(head):
    turtle = head
    rabbit = head
    
    while rabbit:
        if rabbit.next:
            rabbit = rabbit.next.next
            turtle = turtle.next
        else:
            return False
        
        if rabbit is turtle:
            return True
    return False    

File: 9.DetectingCycle.js
//Time O(n) space O(n)
const detectingCycle = (head) => {
    const seen = []
    let currentNode = head
    let hasCycle = false

    while(!seen.includes(currentNode)){
        seen.push(currentNode)
        currentNode = currentNode.next

        if(seen.includes(currentNode)){
            hasCycle = true 
        }
    }

    return hasCycle
}


//Time O(n) Space O(1)
const tortoiseAndHare = (head) => {
    let slow = head;
    let fast = head;
     
     while(fast){
         if(!fast.next) {
             return false;
         } else{
             fast = fast.next.next;
             slow = slow.next;
         }
         
         if(slow == fast){
             return true;
         }
     }
     return false;

}